public/components/map_container/map_container.tsx

```diff
@@ -173,7 +173,7 @@ export const MapContainer = ({
     if (isUpdatingLayerRender || isReadOnlyMode) {
       if (selectedLayerConfig) {
         if (baseLayerTypeLookup[selectedLayerConfig.type]) {
-          handleBaseLayerRender(selectedLayerConfig, maplibreRef);
+          handleBaseLayerRender(selectedLayerConfig, maplibreRef, services);
```

Comment:
I feel it's a bit overwhelming to pass the entire `services` object deeply down, what about use a callback? like:

```typescript

class RegionConflictError extends Error {
  constructor(message) {
    super(message);
    this.name = "RegionConflictError";
  }
}

function onError(e: unknown) {
  if (e instanceof RegionConflictError) {
    services.toastNotifications.addWarning(RegionConflictWarningMsg)
  }
}

// throw RegionConflictError at some point
handleBaseLayerRender(selectedLayerConfig, maplibreRef, onError);
```

Reply:
Got it. But, if we define this class and function then we need to define them outside MapContainer constant., Then, how do we pass `MapServices` to `onError` function?

Reply:
Yes, then it will become like a local function and I think we can't import it in OSMLayerLayerFunctions.ts to call back from there

Reply:
We don't need to import it but pass `onError` as a callback function, like this:
```typescript
export const handleBaseLayerRender = (
  layer: MapLayerSpecification,
  maplibreRef: MaplibreRef,
  onError: (e: unknown) => void
): void => {
  layersFunctionMap[layer.type].render(maplibreRef, layer, onError);
};

export const OSMLayerFunctions = {
  render: (
    maplibreRef: MaplibreRef,
    layerConfig: OSMLayerSpecification,
    onError?: (e: unknown) => void
  ) => {
    // If layer already exist in maplibre source, update layer config
    // else add new layer.
    return hasLayer(maplibreRef.current!, layerConfig.id)
      ? updateLayerConfig(layerConfig, maplibreRef)
      : addNewLayer(layerConfig, maplibreRef, onError);
  },
};

const addNewLayer = (
  layerConfig: OSMLayerSpecification,
  maplibreRef: MaplibreRef,
  onError?: (e: unknown) => void
) => {
  if (maplibreRef.current) {
    const maplibre = maplibreRef.current;
    const { id, source, style } = layerConfig;
    addOSMLayerSource(maplibre, id, source.dataURL);
    fetchStyleLayers(style?.styleURL)
      .then((styleLayers: LayerSpecification[]) => {
        styleLayers.forEach((layer) => {
          const styleLayer = getOSMStyleLayerWithMapLayerId(id, layer);
          addOSMStyleLayer(maplibre, layerConfig, styleLayer);
          setLanguage(maplibreRef, styleLayer);
        });
      })
      .catch((e) => {
        if (onError) {
          onError(e);
        }
      });
  }
};

const fetchStyleLayers = (url: string) => {
  return fetch(url)
    .then((res) => res.json())
    .then((json) => json.layers).catch((e) => {
      if (e instanceof Error) {
        if (e.message === 'NEED TO CHECK IF THIS IS AN ERROR OF REGION CONFLICT')
        throw new RegionConflictError(e.message);
      }
    });
};
```

Reply:
Initially, I thought we need to define both the class and Error function either inside or outside of MapContainer constant. This makes sense. Thanks!

Everything looks good. But, we can't classify if the error is due to region conflict or something else because the error message doesn't share those details.